### PR TITLE
Synthetic Read Aloud

### PR DESCRIPTION
**Pull request by patrickvenetz**
Each Publikator commit can contain multiple "derivatives". GraphQL schema describes it as `Commit.derivatives: [Derivative!]`.

A "derivative" is, no surprise here, "derived" from a commit and might be an audio or a PDF file.

A derivative may be generated asynchronously and wanders through a life cylce: `Pending`, `Failure`, `Ready` and `Detroyed`.

- `Pending` indicates, that a derivative is not ready yet but likely being generated
- `Failure` is a state, describing that a derivative generation failed miserably
- `Ready` signals a derivative being ready to be served (or eaten, or drunken, …)
- `Destroyed` shows destroyed, exploded, vanished, gone for good derivative

In this Pull Request, we introduce `SyntheticReadAload` derivatives, eventually pointing to a computer-generated audio version of commit, read aloud by a synthetic voice.

**Flow**

A derivative is generated either on publish or with mutation `generateDerivative`.

It posts an Mdast document and a webhook URL to Republik TTS wile creating a deriative with status `Pending`. Once Republik TTS is done, webhook URL is called and derivative status set to `Ready` unless an error occured. Later will update derviative status to `Failure`.

`Document.meta.audioSource` returns a URL to an audio file if a user is logged in using an available derivative and `audioSource` in meta is not otherwise set (e.g. uploaded podcast mp3 file).

**Frontend**

If a user can freely access document (and has editor role during an internal testing phase), if a synthetic read aloud version is availble, a box appears underneath an articles action bar.

Bildschirmfoto 2022-03-18 um 10 50 37

Click on large icon will play audio version.

**Publikator**

Version tree allows to generate each commit. It will then post rendered document from commit to Republik TTS server which will eventually report back with generated audio file.

Bildschirmfoto 2022-03-18 um 10 43 33

An pretty much the same for published versions. Once an audio version is generated, you can listen to it in Publikator.

Bildschirmfoto 2022-03-18 um 10 43 12

In case something goes abysmally wrong, one can see that, too. A click onto icon button will reveal details and offer to give it another go.

Bildschirmfoto 2022-03-18 um 10 47 22

Bildschirmfoto 2022-03-18 um 10 48 59

One may suppress syntehtic read aloud per article, setting meta flag `suppressSyntheticReadAloud` – or as a cute, tiny toggle.

Bildschirmfoto 2022-03-15 um 10 51 14

**GraphQL Context `scope`**

Loaders cache their data per request. This is helpful with for GraphQL requests but not so much for context which are kept for a long time e.g. websocket connections, schedulers or Express middlewares.

Introducing `scope` to each context enables us to [disable loaders cache](https://github.com/republik/plattform/pull/46/files#diff-2255c491749b5ebc55ae1a6b0f9d48fda817d1c8862d9378155d52f6c434fdaaR28) if not used in a request basis.

**Uncomitted Changes Route**

`/uncommittedChanges` is deprecated, use namespaced route `/publikator/uncomittedChanges` instead.

**Development**

Find or start a [Republik TTS](https://github.com/republik/tts), a text-to-speech server, and update apps/api/.env:

```
# Text-to-speech server.

# Renders documents as synthetic read aloud audio files.
TTS_SERVER_BASE_URL=https://tts.domain.tld

# Shared secret to sign server requests.
TTS_SIGNATURE_SECRET=some-secret-no-one-knows
```

Make sure, the "Synthetic Read Aloud Server" (S.R.A.S.) can post to your server. If your system is not reachble by S.R.A.S via HTTP(S), you may want to consider setting up a tunnel (e.g. [localtunnel](https://theboroer.github.io/localtunnel-www/) or ngrok if you got the bucks).

While add it, update endpoint for uncommitted changes, too (it got namespaced with /publikator/):

```
API_UNCOMMITTED_CHANGES_URL=http://localhost:5010/publikator/uncommittedChanges
```

Run migrations:

```
yarn migrate:up
```

### CODE REVIEW

**Comment by ovbm** [comment]
frontend implementation:
A section below the action bar. Only visible if audioSource.kind === syntheticReadAloud
The play button is omitted in the Feed so as to not confuse users who expect non-synthetic audio from the current play buttons:
Screenshot 2022-02-24 at 11 15 29
Screenshot 2022-02-24 at 11 27 49
Screenshot 2022-02-24 at 11 15 22

**Comment by ovbm** [comment]
Screenshot 2022-02-24 at 12 13 23
Publikator Buttons

File: apps/www/components/Article/SyntheticAudio.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  }),
+  title: css({
+    ...fontStyles.sansSerifMedium15,
+    margin: '0px 0 4px 0',
+    padding: 0,
+  }),
+  lead: css({ ...fontStyles.sansSerifRegular15, margin: 0, padding: 0 }),
+}
+
+const HeadPhonesIcon = () => (
```
Comment (ovbm) [inline apps/www/components/Article/SyntheticAudio.tsx:71]: not using this anymore, will delete

File: apps/www/components/Article/SyntheticAudio.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            toggleAudioPlayer({
+              audioSource: meta.audioSource,
+              title: meta.title,
+              path: meta.path,
+            })
+          }}
+        />
+        <div {...styles.textContainer}>
+          <p {...styles.title} {...colorScheme.set('fill', 'text')}>
+            Beitrag von synthetischen Stimme vorlesen lassen{' '}
```
Comment (ovbm) [inline apps/www/components/Article/SyntheticAudio.tsx:111]: Copy in translation string

**Review by ovbm — CHANGES_REQUESTED** [review]

**Comment by ovbm** [comment]
frontend self-review done and some changes made. What's the plan for the full review?

**Comment by patrickvenetz** [comment]
@fuenkchen will help with API. Not sure how to manage Publikator and Frontend.

File: packages/backend-modules/translate/translations.json [hunk 3]
```diff
@@ -2235,6 +2235,18 @@
...
       "key": "api/publikator/memo/notPublished/author/email",
       "value": "zurückgezogen"
+    },
+    {
+      "key": "api/publikator/generateDerivative/commit/404",
+      "value": "Commit nicht gefunden"
+    },
+    {
+      "key": "api/publikator/generateDerivative/canNot",
+      "value": "Commit darf nicht als Derivat generiert werden"
```
Comment (fuenkchen) [inline packages/backend-modules/translate/translations.json:2245]: This message will be exposed to end users in Publikator, correct? It's a bit cryptic imho. (Same with value below)

**Review by fuenkchen — COMMENTED** [review]
API part is looking good to me. Just a few comments/hints, maybe they are getting obsolete anyways with fine tuning while testing internally.

File: packages/backend-modules/types/index.d.ts [hunk 4]
```diff
@@ -10,10 +10,12 @@ export interface GraphqlContext extends ConnectionContext {
   t: any
   signInHooks: any
   mail: any
+  scope?: 'request' | 'socket' | 'middleware' | 'master'
```
[the remark this answers is not in the document]
Comment (patrickvenetz) [reply packages/backend-modules/types/index.d.ts:13]: Going for `scheduler`.

**Review by fuenkchen — APPROVED** [review]
API part approved

**Review by annatraussnig — COMMENTED** [review]
The Publikator part looks good at first glance but I haven't got much understanding of the project. It's very standard React code, so I trust @ovbm's review, who knows the project requirements much better than I do.

**Review by ovbm — APPROVED** [review]

**Pull request merged**